File: src/test/java/no/nav/samordning/innlasting/KafkaTestEnvironment.java
package no.nav.samordning.innlasting;

import io.confluent.kafka.serializers.KafkaAvroSerializer;
import no.nav.common.KafkaEnvironment;
import no.nav.samordning.schema.SamordningHendelse;
import org.apache.kafka.clients.producer.KafkaProducer;
import org.apache.kafka.clients.producer.Producer;
import org.apache.kafka.clients.producer.ProducerRecord;
import org.apache.kafka.common.serialization.StringSerializer;

import java.util.*;

import static io.confluent.kafka.serializers.AbstractKafkaAvroSerDeConfig.SCHEMA_REGISTRY_URL_CONFIG;
import static java.lang.Integer.MAX_VALUE;
import static java.util.Collections.emptyList;
import static java.util.Objects.requireNonNull;
import static no.nav.samordning.innlasting.KafkaConfiguration.SAMORDNING_HENDELSE_TOPIC;
import static org.apache.kafka.clients.CommonClientConfigs.BOOTSTRAP_SERVERS_CONFIG;
import static org.apache.kafka.clients.producer.ProducerConfig.*;

class KafkaTestEnvironment {
    private static final int NUMBER_OF_BROKERS = 1;
    private static final String KAFKA_USERNAME = "srvTest";
    private static final String KAFKA_PASSWORD = "opensourcedPassword";
    private static final String TOPIC_NAME = SAMORDNING_HENDELSE_TOPIC;
    private static final List<String> TOPICS = List.of(TOPIC_NAME);

    private static KafkaEnvironment kafkaEnvironment;
    static KafkaConfiguration kafkaConfiguration;
    private static Producer<String, SamordningHendelse> testProducer;

    static void setup() {
        kafkaEnvironment = new KafkaEnvironment(NUMBER_OF_BROKERS, TOPICS, emptyList(), true, false, emptyList(), true, new Properties());
        kafkaConfiguration = new KafkaConfiguration(testEnvironment());
        testProducer = testProducer();
    }

    private static String getUrl() {
        return requireNonNull(kafkaEnvironment.getSchemaRegistry()).getUrl();
    }

    private static String getBrokersURL() {
        return kafkaEnvironment.getBrokersURL();
    }

    private static Map<String, String> testEnvironment() {
        var testEnvironment = new HashMap<String, String>();
        testEnvironment.put("KAFKA_BOOTSTRAP_SERVERS", getBrokersURL());
        testEnvironment.put("SCHEMA_REGISTRY_URL", getUrl());
        testEnvironment.put("KAFKA_USERNAME", KAFKA_USERNAME);
        testEnvironment.put("KAFKA_PASSWORD", KAFKA_PASSWORD);
        testEnvironment.put("KAFKA_SASL_MECHANISM", "PLAIN");
        testEnvironment.put("KAFKA_SECURITY_PROTOCOL", "PLAINTEXT");
        return testEnvironment;
    }

    private static Producer<String, SamordningHendelse> testProducer() {
        var producerProperties = new Properties();
        producerProperties.put(BOOTSTRAP_SERVERS_CONFIG, getBrokersURL());
        producerProperties.put(SCHEMA_REGISTRY_URL_CONFIG, getUrl());
        producerProperties.put(KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
        producerProperties.put(VALUE_SERIALIZER_CLASS_CONFIG, KafkaAvroSerializer.class);
        producerProperties.put(ACKS_CONFIG, "all");
        producerProperties.put(RETRIES_CONFIG, MAX_VALUE);
        return new KafkaProducer<>(producerProperties);
    }

    static void populate_hendelse_topic(String TPNR, SamordningHendelse samordningHendelse) {
        var record = new ProducerRecord<>(TOPIC_NAME, TPNR, samordningHendelse);
        testProducer.send(record);
        testProducer.flush();
    }
}

File: src/test/kotlin/no/nav/samordning/innlasting/KafkaTestEnvironment.kt
package no.nav.samordning.innlasting

import io.confluent.kafka.serializers.AbstractKafkaAvroSerDeConfig.SCHEMA_REGISTRY_URL_CONFIG
import io.confluent.kafka.serializers.KafkaAvroSerializer
import no.nav.common.JAASCredential
import no.nav.common.KafkaEnvironment
import no.nav.common.embeddedutils.ServerBase
import no.nav.samordning.innlasting.KafkaConfiguration.Companion.SAMORDNING_HENDELSE_TOPIC
import no.nav.samordning.schema.SamordningHendelse
import org.apache.kafka.clients.CommonClientConfigs.BOOTSTRAP_SERVERS_CONFIG
import org.apache.kafka.clients.producer.KafkaProducer
import org.apache.kafka.clients.producer.Producer
import org.apache.kafka.clients.producer.ProducerConfig.*
import org.apache.kafka.clients.producer.ProducerRecord
import org.apache.kafka.common.serialization.StringSerializer
import java.lang.Integer.MAX_VALUE
import java.util.*
import java.util.Collections.emptyList
import java.util.Objects.requireNonNull

internal object KafkaTestEnvironment {
    private const val NUMBER_OF_BROKERS = 1
    private const val KAFKA_USERNAME = "srvTest"
    private const val KAFKA_PASSWORD = "opensourcedPassword"
    private const val TOPIC_NAME = SAMORDNING_HENDELSE_TOPIC
    private val TOPICS = listOf(TOPIC_NAME)

    private lateinit var kafkaEnvironment: KafkaEnvironment
    lateinit var kafkaConfiguration: KafkaConfiguration
    private lateinit var testProducer: Producer<String, SamordningHendelse>

    private val url: String
        get() = requireNonNull<ServerBase>(kafkaEnvironment.schemaRegistry).url

    private val brokersURL: String
        get() = kafkaEnvironment.brokersURL

    fun setupKafkaEnvironment() {
        kafkaEnvironment = KafkaEnvironment(NUMBER_OF_BROKERS, TOPICS, emptyList<KafkaEnvironment.TopicInfo>(), withSchemaRegistry = true, withSecurity = false, users = emptyList<JAASCredential>(), autoStart = true, brokerConfigOverrides = Properties())
        kafkaConfiguration = KafkaConfiguration(testEnvironment())
        testProducer = testProducer()
    }

    private fun testEnvironment(): Map<String, String> {
        return mapOf(
                "KAFKA_BOOTSTRAP_SERVERS" to brokersURL,
                "SCHEMA_REGISTRY_URL" to url,
                "KAFKA_USERNAME" to KAFKA_USERNAME,
                "KAFKA_PASSWORD" to KAFKA_PASSWORD,
                "KAFKA_SASL_MECHANISM" to "PLAIN",
                "KAFKA_SECURITY_PROTOCOL" to "PLAINTEXT"
        )
    }

    private fun testProducer() = KafkaProducer<String, SamordningHendelse>(
            Properties().apply {
                put(BOOTSTRAP_SERVERS_CONFIG, brokersURL)
                put(SCHEMA_REGISTRY_URL_CONFIG, url)
                put(KEY_SERIALIZER_CLASS_CONFIG, StringSerializer::class.java)
                put(VALUE_SERIALIZER_CLASS_CONFIG, KafkaAvroSerializer::class.java)
                put(ACKS_CONFIG, "all")
                put(RETRIES_CONFIG, MAX_VALUE)
            })

    fun populate_hendelse_topic(TPNR: String, samordningHendelse: SamordningHendelse) {
        val record = ProducerRecord(TOPIC_NAME, TPNR, samordningHendelse)
        testProducer.send(record)
        testProducer.flush()
    }
}
